### DISCUSSION TITLE
Trouble using addCollection

### DISCUSSION THREAD

**Discussion by trinkel** [11ty/eleventy#2412] [Q&A]
I am having problems adding a collection from json files but as far as I can tell, the collection is empty. This is the file tree:

```
myproject
| src
| | _includes
| | | layouts
| | | | base.njk
| | _competitions
| | | eauclaire.json
| | | hiawathaland.json
| | | maplewood.json
| | | mids.json
| index.njk
| .eleventy.js
```

I have the following code:

```
// .eleventy.js
eleventyConfig.addCollection("competitions", (collection) => {
	return collection.getFilteredByGlob("./src/competitions/*.json");
});
```

```
// index.njk
---
layout: layouts/base.njk
title: Competition Results
---
```

```
// base.njk
// (log is a filter that calls console.log)

{{ collections | log }}
```

Sample contents of a json file:
```
{
	"name": "Eau Claire Figure Skating Competition",
	"date": "2022-03-11",
	"dateText": "March 11–13, 2022",
	"location": "Eau Claire, WI",
	"teams": [
		{
			"name": "Aspire Pre-Preliminary",
			"place": "3rd Place",
			"score": ""
		},
		{
			"name": "Open Juvenile",
			"place": "Exhibition",
			"score": ""
		},
		{
			"name": "Pre Juvenile",
			"place": "Exhibition",
			"score": ""
		}
	]
}
```

Logged output:
```
{
  all: [
    {
      template: [Template],
      inputPath: './src/index.njk',
      fileSlug: '',
      filePathStem: '/index',
      data: [Object],
      date: 2022-05-25T17:53:05.647Z,
      outputPath: 'dist/index.html',
      url: '/',
      checkTemplateContent: true,
      templateContent: [Getter/Setter],
      _templateContent: '\n'
    }
  ],
  competitions: []
}
```

Note that addCollection creates `competitions`, but it is empty. Pointing addCollection directly to a json file doesn't change anything.

At this point I am not trying to use the data, I am just trying to collect it. If you need more info, I'm happy to supply it.

So, am I doing it wrong or am I trying to do something that won't work in the first place?

**Comment by pdehaan** [comment]
Per https://www.11ty.dev/docs/collections/#getfilteredbyglob(-glob-),
> _"Note: `getFilteredByGlob` will not "find" files that are not supported by Eleventy. For example, a file with the extension `.ray` will be ignored even if it would match the glob."_

So I think that's what you're seeing here. Because 11ty doesn't know how to process a ".json" file, it just ignores them. I was able to get this working by creating a custom template format for .json files, but not sure if it's 100% correct:

https://www.11ty.dev/docs/languages/custom/

```js
  eleventyConfig.addTemplateFormats("json");
  eleventyConfig.addExtension("json", {
    async compile(inputContent) {
      const output = JSON.parse(inputContent);
      return async () => {
        return JSON.stringify(output);
      };
    }
  });
```

But now I can create a src/competitions/index.njk with the following:

```njk
{% for comp in collections.competitions %}
<div>
{{ comp.data.name | safe }}
</div>
{% endfor %}
```

And it seems to log out the expected data. But on the downside, now the .json files are also written into my _site dist folder, which may not be what you want.

I'm not completely certain what you are wanting with the .json files, or why they're in a custom collection versus putting them in your _data/ directory and treating that like a faux-collection. If you just want a list of filenames or filenames and values, we might be able to do the glob outside of `.addCollection()` using **fast-glob** or something else and then return an array of values.

**Reply by trinkel** [reply]
Hmmmm. Yes, I've read that paragraph several times trying to figure this out.   I was thinking json was a supported file type. That makes my troubleshooting results make a lot more sense. I'll give this a shot when I get home. 

I take it json is ok in global data but not collections?

Thanks!

**Reply by trinkel** [reply]
@pdehaan: Missed one of your questions . . .

Basically I'm building a competition results page. It will have tables for several years worth of competition results for several teams done as a layout. I took it out of `_data` cuz I really only need it for one page, not so much as global data (I'm thinking performance here, but I'm not sure if that's necessary). I also want to iterate over the years and the teams. From my limited knowledge at this point I thought it may be easier to organize and create data as a custom collection.

**Reply by pdehaan** [reply]
@trinkel, another proposal:

(EDIT: See https://github.com/11ty/eleventy/discussions/2412#discussioncomment-2840801 below)

**Reply by trinkel** [reply]
@pdehaan: Dude! Thanks! Totally on point. Actually this is really the organization I was trying to accomplish in the first place. I will flesh this out in my project.

I tried the stuff you mentioned in your first answer and still came up with empty data. I'm sure I'm just missing a piece somewhere. I'd still like to figure it out, but then again, meh. If I ignore it for awhile, maybe it will come to me in the middle of the night.

As a relative noob to discussions . . . I want to mark these replies as the answer because the first answer wasn't actually the answer. But it looks like I can't do that because they're "replies" not "answers"? Is there a way to break a reply out to an answer to mark it or do I just mark your first answer? What's the protocol?

Thanks again.

**Reply by pdehaan** [reply]
Ah, yeah. I guess I screwed that up. 
~if you want I can copy/paste my second proposal into a new answer when I get to my laptop.~
DONE: See https://github.com/11ty/eleventy/discussions/2412#discussioncomment-2840801 below

**Comment by pdehaan** [comment]
@trinkel, another proposal:

```sh
src/
├── _includes/
│   └── layouts/
│       └── base.njk
├── competitions/
│   ├── competitions.json
│   ├── eauclaire.11tydata.json
│   ├── eauclaire.njk
│   ├── hiawathaland.11tydata.json
│   ├── hiawathaland.njk
│   ├── index.njk
│   ├── maplewood.11tydata.json
│   ├── maplewood.njk
│   ├── mids.11tydata.json
│   └── mids.njk
└── index.njk

3 directories, 12 files
```

Where my src/competitions/competitions.json directory data file looks like this, and apply the `layout` and `tags` to each file in the subdirectory (which will auto-create the collection for you):
```json
{
  "layout": "layouts/base.njk",
  "tags": ["competitions"]
}
```

And the src/competitions/eauclaire.11tydata.json template data file looks like:

```json
{
	"name": "Eau Claire Figure Skating Competition",
	"date": "2022-03-11",
	"dateText": "March 11–13, 2022",
	"location": "Eau Claire, WI",
	"teams": [
		{
			"name": "Aspire Pre-Preliminary",
			"place": "3rd Place",
			"score": ""
		},
		{
			"name": "Open Juvenile",
			"place": "Exhibition",
			"score": ""
		},
		{
			"name": "Pre Juvenile",
			"place": "Exhibition",
			"score": ""
		}
	]
}
```

And the src/competitions/eauclaire.njk template looks like:
```njk
---
title: Eauclaire
---

<h1>{{ name }}</h1>
<time date="{{ date }}">{{ dateText }}</time>
```

Now instead of global data files you have template data files which will only apply to their specific Nunjucks template.
Or you could convert the template data file to YAML and put it in the eauclaire.njk template front matter, or use JSON front matter.

--- 

### JSON front matter example
src/competitions/hiawathaland.njk
```njk
---json
{
  "name": "Hiawathaland Figure Skating Competition",
  "date": "2022-03-21",
  "dateText": "March 21–30, 2022",
  "location": "Hiawathaland, WI",
  "teams": [
    {
      "name": "Aspire Pre-Preliminary",
      "place": "3rd Place",
      "score": ""
    },
    {
      "name": "Open Juvenile",
      "place": "Exhibition",
      "score": ""
    },
    {
      "name": "Pre Juvenile",
      "place": "Exhibition",
      "score": ""
    }
  ]
}
---

<h1>{{ name }}</h1>
<time date="{{ date }}">{{ dateText }}</time>
```

---

### YAML front matter example
src/competitions/maplewood.njk
```njk
---
name: Maplewood Figure Skating Competition
date: 2022-03-21
dateText: March 21–30, 2022
location: Maplewood, WI
teams:
  - name: Aspire Pre-Preliminary
    place: 3rd Place
    score: ""
  - name: Open Juvenile
    place: Exhibition
    score: ""
  - name: Pre Juvenile
    place: Exhibition
    score: ""
---

<h1>{{ name }}</h1>
<time date="{{ date }}">{{ dateText }}</time>
```

**Comment by pdehaan** [comment]
I know you said you didn't want global data, but for future travellers, here's a solution which uses nested global directories in the _data folder and then paginates over the results (using Nunjucks):

```sh
tree src
src/
├── _data/
│   └── competitions/
│       ├── eauclaire.json
│       └── hiawathaland.json
├── _includes/
│   └── layouts/
│       └── base.njk
├── competitions/
│   ├── index.njk
│   └── pagination.njk
└── index.njk
```

### ./src/_data/competitions/eauclaire.json
```json
{"name":"Eau Claire"}
```

### ./src/competitions/index.njk
```njk
---
title: Competitions
---

<h1>{{ title }}</h1>
{% for dataslug, competition in competitions %}
<article>
  <h1>{{ competition.name }}</h1>
  <p><a href="./{{ competition.name | slugify }}/">Details</a></p>
  <pre>{{ competition | dump | safe }}</pre>
</article>
{% endfor %}
```

### ./src/competitions/pagination.njk
```njk
---
pagination:
  data: competitions
  size: 1
  alias: competition
  resolve: values
layout: layouts/base.njk
permalink: "competitions/{{ competition.name | slugify }}/"
---

<h1>{{ competition.name }}</h1>
```

Not sure that it's any better or worse. Although it is a bit trickier to manage `permalink`s when using fake-collections.
Probably not a lot of benefit here versus using the JSON or YAML front matter approach. Both seem to be reduced to adding one new file in a single location.

**Reply by trinkel** [reply]
Thanks @pdehaan !  This is a lot of good information to chew on. My brain is already off on tangents based on your input.

I am in and out a lot this weekend, but I will play with this as I have time and post back where I land.

**Reply by trinkel** [reply]
Still futzing, but I'm getting all three scenarios (global data, collections, template data) to work and I will report back.

I'm starting to think that I'm over thinking the whole performance thing in deciding how to set this up.

Is the difference between these three data models not so much a question of performance as mechanics: how it's used/accessed (global data and collections) and scoping (template data)? Is there anything detrimental to using any of these methods (performance hit or whatever) or is it really just a matter of where and how the data is exposed?
